pykafka/producer.py

```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
         # only *some* messages when a leader changes. Therefore, we don't want
         # a random partition distribution changing that on the retry.
         self._produce(self._partition_messages(messages), 0)
+
+
+class AsyncProducer(Producer):
+    """
+    This class implements the asynchronous producer logic found in the
+    JVM driver. In inherits from the synchronous implementation.
+    """
+    def __init__(self, *args, **kwargs):
+        """Instantiate a new AsyncProducer
+
+        For argument documentation, see :class:`pykafka.producer.Producer`
+        """
+        super(AsyncProducer, self).__init__(*args, **kwargs)
+        self._setup_workers()
+
+    def _setup_workers(self):
+        """Spawn workers for connected brokers
+
+        Creates one :class:`queue.Queue` instance and one worker thread per
+        broker. The queue holds requests waiting to be sent, and the worker
+        thread reads the end of the queue and sends requests.
+        """
+        self._queues_by_leader = {}
+        for partition in self._topic.partitions.values():
+            if partition.leader.id not in self._queues_by_leader:
+                self._queues_by_leader[partition.leader.id] = self._setup_queue(partition.leader)
+
+    def _setup_queue(self, broker):
+        """Spawn a worker for the given broker
+
+        Creates a :class:`queue.Queue` instance and a worker thread for the
+        given broker.
+
+        :param broker: The broker for which to instantiate this queue and worker
+        :type broker: :class:`pykafka.broker.Broker`
+        """
+        def worker():
+            while True:
+                time.sleep(.0001)
+                try:
+                    request, attempt = request_queue.get_nowait()
+                except Empty:
+                    continue
+                self._send_request(broker, request, attempt)
+        request_queue = Queue()
+        log.info("Starting new produce worker thread for broker %s", broker.id)
+        self._cluster.handler.spawn(worker)
+        return request_queue
+
+    def _request_built_callback(self, leader, request, attempt):
+        """Callback called when a request is ready to be sent
+
+        Enqueues the request in the queue associated with `leader`
+
+        :param leader: The broker to which to send the request
+        :type leader: :class:`pykafka.broker.Broker`
+        :param request: The request to send
+        :type request: :class:`pykafka.protocol.ProduceRequest`
+        :param attempt: The attempt number for this send attempt
+        :type attempt: int
+        """
+        self._queues_by_leader[leader.id].put((request, attempt))
```

Comment:
That sounds useful. I think the java producer has a `queue.buffering.max.messages` config parameter for that - librdkafka has the same.